awx/main/utils/external_logging.py

```diff
@@ -0,0 +1,70 @@
+import urllib.parse as urlparse
+
+from django.conf import settings
+
+from awx.main.utils.reload import supervisor_service_command
+
+
+def reconfigure_rsyslog():
+    tmpl = ''
+    parts = ['$IncludeConfig /etc/rsyslog.conf']
+    if settings.LOG_AGGREGATOR_ENABLED:
```

Comment:
This function would be really easy (and fun) to unit test. You don't come by those every day. All it would take would be to move it to a helper.

```python
def get_rsyslog_template(settings_vals):
    ... # you want to test this stuff
    return tmpl

def reconfigure_rsyslog():
    # inside reconfigure_rsyslog
    tmpl = get_rsyslog_template(settings)
    with open('/var/lib/awx/rsyslog.conf', 'w') as f:
        f.write(tmpl + '\n')
    supervisor_service_command(command='restart', service='awx-rsyslogd')
```
At that point, you're just building strings based on input values to `get_rsyslog_template`. You could shred through tons of permutations and expected output strings strings using `@parametrize` and it would be fast.

If you don't find at least one bug doing this, I would be surprised.

Reply:
I agree; we should **definitely** refactor this a bit and extensively unit test what gets generated based on the values of `settings`.